modules/ding_place2book/ding_place2book.module

```diff
@@ -353,13 +353,26 @@ function ding_place2book_get_ticket_link($node, $place2book_id = NULL) {
 
     $status_type = _ding_place2book_status_processing($p2b_result);
 
-    $node->field_ding_event_ticket_link[LANGUAGE_NONE][0]['url'] = $status_type['url'];
-    $node->field_ding_event_ticket_link[LANGUAGE_NONE][0]['place2book_id'] = $place2book_id;
-    $node->field_ding_event_ticket_link[LANGUAGE_NONE][0]['status'] = $status_type['type'];
-    field_attach_presave('node', $node);
-    field_attach_update('node', $node);
-    // Clear the static loading cache.
-    entity_get_controller('node')->resetCache(array($node->nid));
+    $lock_id = implode('|', [
+      'entity:' => 'node',
+      'nid:' => $node->nid,
+      'type:' => $node->type,
+    ]);
+
+    // Handle concurrency.
+    // Field save might deliver integrity constraint exceptions when invoked
+    // by numerous simultaneous requests.
+    if (lock_acquire($lock_id, 1)) {
```

Comment:
The need for this change is caused by the change in #1104 where `ding_place2book_get_ticket_link()` is called from a hook `node_view()` implemenetation.

@guddo @bozeak: Can  provide a more detailed explanation of why we need to retrieve the ticket link from place2book on each node view? I would think it to be static once the event is created.

*If* we need to retrieve the link dynamically then we should refactor the field into using the `virtual_field` module like we do for similar fields instead of trying to store something in the database where it is not needed.

Reply:
As written in the issue, this was done in PR#1096 and PR#1104, due to the lack of older nodes not getting the correct fields changed. It was then moved to node_view() to make sure that older nodes also got this change. However, in hindsight, this is not a feasible implementation as frontend users should not be able to alter anything frankly.

We do need to retrieve the status of the link dynamically, however it should be refactored somehow.

Reply:
> It was then moved to node_view() to make sure that older nodes also got this change. However, in hindsight, this is not a feasible implementation as frontend users should not be able to alter anything frankly.

I agree. If older nodes were not getting the correct fields changed they should have been updated in a `hook_update()`or the like.

Based on your explanation, are you also saying that Place2book is using an existing field `field_ding_event_ticket_link` which is a field that editors can use to manually specify a link for an event. If this is the case then this is obviously not something we can move to `virtual_field`.

If reacting to node changes on the site (`hook_node_insert/update()`) is not sufficient could we then update the link in the background using a cron job or the like?

Do you have other suggestions regarding how this should be handled in a better way?